src/backend/EdFi.DataManagementService.Backend.Postgresql/Operation/SqlAction.cs

```diff
@@ -0,0 +1,192 @@
+// See the LICENSE and NOTICES files in the project root for more information.
+
+using System.Text.Json;
+using EdFi.DataManagementService.Backend.Postgresql.Model;
+using EdFi.DataManagementService.Core.External.Model;
+using Npgsql;
+
+namespace EdFi.DataManagementService.Backend.Postgresql.Operation;
+
+/// <summary>
+/// A facade of all the DB interactions. Any action requiring SQL statement execution should be here.
+/// Connections and transactions are managed by the caller.
+/// Exceptions are handled by the caller.
+/// </summary>
+public interface ISqlAction
+{
+    public Task<Document?> FindDocumentByReferentialId(
+        ReferentialId referentialId,
+        PartitionKey partitionKey,
+        NpgsqlConnection connection,
+        NpgsqlTransaction transaction
+    );
+
+    public Task<long> InsertDocument(
+        Document document,
+        NpgsqlConnection connection,
+        NpgsqlTransaction transaction
+    );
+
+    public Task<long> InsertAlias(Alias alias, NpgsqlConnection connection, NpgsqlTransaction transaction);
+
+    public Task<int> UpdateDocumentEdFiDoc(
+        int documentPartitionKey,
+        Guid documentUuid,
+        JsonElement edfiDoc,
+        NpgsqlConnection connection,
+        NpgsqlTransaction transaction
+    );
+}
+
+public record UpsertDocumentSqlResult(bool Inserted, long DocumentId);
+
+/// <summary>
+/// A facade of all the DB interactions. Any action requiring SQL statement execution should be here.
+/// Connections and transactions are managed by the caller.
+/// Exceptions are handled by the caller.
+/// </summary>
+public class SqlAction : ISqlAction
+{
+    /// <summary>
+    /// Returns a single Document from the database corresponding to the given ReferentialId,
+    /// or null if no matching Document was found.
+    /// </summary>
+    public async Task<Document?> FindDocumentByReferentialId(
+        ReferentialId referentialId,
+        PartitionKey partitionKey,
+        NpgsqlConnection connection,
+        NpgsqlTransaction transaction
+    )
+    {
+        await using NpgsqlCommand command =
+            new(
+                @"SELECT * FROM public.Documents d
+                INNER JOIN public.Aliases a ON a.DocumentId = d.Id AND a.DocumentPartitionKey = d.DocumentPartitionKey
+                WHERE a.ReferentialPartitionKey = $1 AND a.ReferentialId = $2;",
+                connection,
+                transaction
+            )
+            {
+                Parameters =
+                {
+                    new() { Value = partitionKey.Value },
+                    new() { Value = referentialId.Value },
+                }
+            };
+
+        await using NpgsqlDataReader reader = await command.ExecuteReaderAsync();
+
+        if (!reader.HasRows)
+        {
+            return null;
+        }
+
+        // Assumes only one row returned (should never be more due to DB unique constraint)
+        await reader.ReadAsync();
+
+        return new(
```

Comment:
Not using Dapper here because we can't use prepared statements with it for our three main tables.